File: 2021/day_07/advent_07.py
#!/usr/bin/python3

import sys
from collections import defaultdict

cost_increment = int(sys.argv[1] or 0)

input = sys.stdin.read()
pos = sorted([int(it) for it in input.split(',')])
idx = defaultdict(int)
for p in pos:
    idx[p] += 1


def cost_to(dest):
    cost = 0
    for p, amount in idx.items():
        length = abs(p - dest)
        each_cost = int(length * (2 + cost_increment * (length - 1)) / 2)
        group_cost = amount * each_cost
        cost += group_cost
    return cost


min_cost = 1e16
lo, hi = 0, len(pos)
delta = None

while lo < hi:
    mid = lo + ((hi - lo) // 2)
    mid_left = mid  # looking for first distinct dest to the left
    while pos[mid_left] == pos[mid]:
        mid_left -= 1

    mid_cost = cost_to(pos[mid])
    if min_cost > mid_cost:
        destination = pos[mid]
        min_cost = mid_cost

    delta = mid_cost - cost_to(pos[mid_left])
    hi = mid if delta >= 0 else hi
    lo = mid + 1 if delta < 0 else lo

print(f'Destination {destination}\nCost {min_cost}')

File: 2021/day_07/advent_07.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

func main() {
	costIncrement := readParam()
	var pos = readInts()
	sort.Ints(pos)
	idx := map[int]int64{}
	for i := 0; i < len(pos); i++ {
		idx[pos[i]] += 1
	}
	minCost := int64(1e16)
	for dest := pos[0]; dest < pos[len(pos)-1]+1; dest++ {
		cost := int64(0)
		idx_keys := sortedKeys(idx)
		for i := 0; i < len(idx_keys); i++ {
			p := idx_keys[i]
			amount := idx[p]
			length := abs(p - dest)
			eachCost := int64(length * (2 + costIncrement*(length-1)) / 2)
			groupCost := amount * eachCost
			cost += groupCost
		}
		if int64(cost) > minCost {
			fmt.Printf("Destination %d\nCost %d\n", dest, minCost)
			os.Exit(0)
		}
		minCost = cost
	}
	log.Fatalf("No solution found. minCost so far: %d\n", minCost)
}

func readParam() int {
	value, err := strconv.Atoi(os.Args[1])
	if err != nil {
		return 0
	}
	return value
}

func readInts() []int {
	reader := bufio.NewReader(os.Stdin)
	str, _ := reader.ReadString('\n')

	var ints []int
	err := json.Unmarshal([]byte(fmt.Sprintf("[%s]", str)), &ints)
	if err != nil {
		log.Fatalf("FATAL: %s", err)
	}
	return ints
}

func sortedKeys(m map[int]int64) []int {
	keys := make([]int, len(m))
	i := 0
	for k := range m {
		keys[i] = k
		i++
	}
	sort.Ints(keys)
	return keys
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
